modify nft

akayi07 opened the pull request.

yymone commented on pallets/swap/src/lib.rs:
```diff
@@ -610,14 +605,13 @@ decl_module! {
             origin,
             sign: [u8; 64],
-            account_index: AccountIndex,
             nft_id: NFTId,
```
suggest put this arg after account_index to follow event args

yymone commented on pallets/swap/src/lib.rs:
```diff
@@ -675,8 +668,8 @@ decl_module! {
 
-            nft_widthdraw::<T>(&account_index, &nft_id);
-            NonceMap::<T>::insert(&account, new_nonce);
+            nft_withdraw::<T>(&account_index, &nft_id)?;
```
We should making the treating steps consistant. 
This is the doing work function.
In DepositNFT, it is:
  nft_add::<T>(&account_index, &nft_id)?;
  let op = Ops::DepositNFT
  ..
  Self::deposit_event

Thus we should keep such order.

yymone commented on pallets/swap/src/lib.rs:
```diff
@@ -691,42 +684,122 @@ decl_module! {
             let req_id = req_id_get::<T>()?;
-            if recipent >= AccountIndexCount::get() {
+            if recipient >= AccountIndexCount::get() {
                 return Err(Error::<T>::InvalidAccount)?;
```
This is a good check.
However, if you check all other nft function's, I can see the code do not check such things for all essential args.
For example:
deposit_nft's accountIndex, nftid, etc.

Please consider all nft function's user-input args to chech whether they exceed ranges. (origin do not need it as it is from our B/E system)

Also it may be worth to have some help function to do such things.

yymone commented on pallets/swap/src/lib.rs:
```diff
@@ -691,42 +684,122 @@ decl_module! {
 
-            nft_transfer::<T>(&account_index, &recipent, &nft_id);
-            NonceMap::<T>::insert(&account, new_nonce);
+            nft_transfer::<T>(&account_index, &recipient, &nft_id)?;
```
We should keep consistent, either use this order, or use deposit order.

yymone commented on pallets/swap/src/aux.rs:
```diff
@@ -198,7 +198,7 @@ pub fn nft_add<T: Config>(
 }
 
-pub fn nft_widthdraw<T: Config>(
+pub fn nft_withdraw<T: Config>(
```
Need check nft empty

yymone replied:
Also, the implementation is wrong.
Xin just did a framework and the content is not fully filled.
For withDraw, if you check the ts file in ZKP repo , you can see it also need to do:

    // STEP2: if dataPath[2]'s leafValues[1] != 0, return biddingAmount to bidder
    // circuits: check balance of current bidder dosen't overflow(< 2 ^ 250)

akayi07 replied:
I think there's no need to check nft empty because in aux.rs line 212, nft has been set to empty.

akayi07 replied:
I will check nft empty in unit test

yymone commented on pallets/swap/src/aux.rs:
```diff
@@ -216,8 +216,29 @@ pub fn nft_transfer<T: Config>(
 ) -> Result<(), Error<T>> {
     let nft = NFTMap::get(&nft_id);
     nft.checked_owner(from_index)?;
+    NFTMap::insert(nft_id, (to_index, nft.1, nft.2));
```
orignal code is "NFTMap::insert(nft_id, (to_index, U256::from(0), bidder));"
why change to nft.1 nft.2?

yymone commented on pallets/swap/src/aux.rs:
```diff
@@ -216,8 +216,29 @@ pub fn nft_transfer<T: Config>(
+    amount: Amount,
+    nft_id: &NFTId,
+) -> Result<(), Error<T>> {
+    let nft = NFTMap::get(&nft_id);
```
check nft exist

yymone commented on pallets/swap/src/aux.rs:
```diff
@@ -216,8 +216,29 @@ pub fn nft_transfer<T: Config>(
+    account_index: &AccountIndex,
+    nft_id: &NFTId,
+) -> Result<(), Error<T>> {
+    let nft = NFTMap::get(&nft_id);
```
check nft exist.
check nft.2 exist.

akayi07 replied:
line 256 has checked nft exist

yymone commented on pallets/swap/src/aux.rs:
```diff
@@ -216,8 +216,29 @@ pub fn nft_transfer<T: Config>(
+    return Ok(());
+}
+
+pub fn nft_finalize<T: Config>(
```
Miss functional:
    // STEP2: update balance of current owner
    // circuits: check balance dosen't overflow

yymone commented on pallets/swap/src/aux.rs:
```diff
@@ -216,8 +216,29 @@ pub fn nft_transfer<T: Config>(
+    return Ok(());
+}
+
+pub fn nft_bid<T: Config>(
```
Miss functional:

    // STEP2: if dataPath[3]'s leafValues[1] !=0, update balance of current accountIndex
    // circuits: check balance dosen't overflow
    // get current accountIndex of nft

    // STEP3: update balance of accountIndex

yymone commented on pallets/swap/src/tests.rs:
```diff
@@ -1,23 +1,340 @@
 
 #[test]
-fn it_works_for_default_value() {
+fn nft_works() {
```
It is good to build up the framework for the unit test.
However, these unit tests are useless.
The purpose of unit test is to test the function works well.
The only assertings in this test file are just like:
    
    assert_ok!(SwapModule::deposit
    assert_ok!(SwapModule::withdraw_nft

This is not very helpful because it will only fail when the function crash (rust panic!)
It does not test any real functionality.
(Also just as what I metioned in the aux.rs file, most nft functions do not implement the whole functionality)

yymone commented on pallets/swap/src/aux.rs:
```diff
@@ -175,12 +175,16 @@ trait NFTData<T: Config> {
     fn checked_empty(&self) -> Result<(), Error<T>> {
-        return Ok(());
+        if self.0 != 0u32 {
+            return Err(Error::<T>::NFTNoAuthority)?;
```
This error looks like not meaningful.
I think it should be InvalidNFTIndex, as it is because we hope it should be 0 but the input accountIndex (for NFT) is invalid.

yymone replied:
And why we need "?" after it?

akayi07 replied:
Yes, it's a mistake to add ? after it.

yymone commented on pallets/swap/src/aux.rs:
```diff
@@ -175,12 +175,16 @@ trait NFTData<T: Config> {
     fn checked_owner(&self, account_index: &AccountIndex) -> Result<(), Error<T>> {
         if self.0 != *account_index {
-            return Err(Error::<T>::InvalidTokenIndex)?;
+            return Err(Error::<T>::IsNotOwner)?;
```
why "?"

akayi07 replied:
Yes, it's a mistake to add ? after it because the function return Result rather than DispatchResult.

yymone commented on pallets/swap/src/aux.rs:
```diff
@@ -216,8 +224,44 @@ pub fn nft_transfer<T: Config>(
+        let new_balance_amount0 = balance_add::<T>(&nft.2.unwrap(), &NFT_TOKEN_INDEX, nft.1)?;
+        balance_set(&nft.2.unwrap(), &NFT_TOKEN_INDEX, new_balance_amount0);
+    }
+    let new_balance_amount1 = balance_sub::<T>(bidder, &NFT_TOKEN_INDEX, amount)?;
```
Here we have problem.
If pervious balance_add and balance_set success and this balance_sub failed and return, the memory status will be incorrect.

I suggest adjust the order and do all balance_set after we call balance_add and balance_sub to avoid such issue.

akayi07 replied:
I understood. If we execute the funtion twice, the balance_add will be executed twice successfully.

yymone commented on pallets/swap/src/aux.rs:
```diff
@@ -216,8 +224,44 @@ pub fn nft_transfer<T: Config>(
+) -> Result<(), Error<T>> {
+    let nft = NFTMap::get(&nft_id);
+    if nft.2 == None {
+        return Err(Error::<T>::NFTNoAuthority)?;
```
I think you can add an new error like InvaidNFTBidder

akayi07 replied:
just like https://github.com/DelphinusLab/delphinus-substrate-node/pull/9#discussion_r846880460, I think it should be InvalidNFTIndex?

yymone commented on pallets/swap/src/lib.rs:
```diff
@@ -13,13 +13,16 @@ mod aux;
 mod types;
 
-use aux::*;
+pub use aux::*;
```
why change this to pub?
This one looks not good as it is doing re-exporting.
If test files need it you can call "use aux::*" in the tests.rs?

akayi07 replied:
Yes, It is a mistake. I do not need pub.

yymone commented on pallets/swap/src/lib.rs:
```diff
@@ -13,13 +13,16 @@ mod aux;
 #[cfg(test)]
 mod mock;
 
+#[cfg(test)]
```
Could you explain why this is added here in lib.rs? I didn't see any unit tests write here.
Is it because of substrate node framework?

akayi07 replied:
I add #[test] in tests.rs, so I should add #[cfg(test)] in lib.rs. Here is the official example: https://github.com/substrate-developer-hub/substrate-node-template/blob/467927bda05a56dfe57690aec93ff504a6009daa/pallets/template/src/lib.rs#L12. Of course we can move unit tests into lib.rs.

yymone commented on pallets/swap/src/lib.rs:
```diff
@@ -616,7 +612,15 @@ decl_module! {
+            let caller_account_index = get_account_index::<T>(&who)?;
+
+            if account_index >= AccountIndexCount::get() || account_index == 0u32 {
+                return Err(Error::<T>::InvalidAccount)?;
```
I don't think we need "?" here. Or you have any comment why we have it here?

yymone commented on pallets/swap/src/lib.rs:
```diff
@@ -616,7 +612,15 @@ decl_module! {
+            }
+
+            if nft_id >= MAX_NFTINDEX_COUNT || nft_id == 0u32 {
+                return Err(Error::<T>::InvalidNFTIndex)?;
```
This need to be a function.
You can write a function named ValidationNFTIndex<T: Config>(nft_id: NFTId) -> Result(_, Error<T>) {
    if nft_id >= MAX_NFTINDEX_COUNT || nft_id == 0u32 {
                return Err(Error::<T>::InvalidNFTIndex);
}

and call it with "?" at end to implement the early return.

akayi07 replied:
I understood. The code appears more than once.

yymone commented on pallets/swap/src/lib.rs:
```diff
@@ -656,32 +661,34 @@ decl_module! {
+            let account_index = get_account_index::<T>(&who)?;
+
+            if nft_id >= MAX_NFTINDEX_COUNT || nft_id == 0u32 {
+                return Err(Error::<T>::InvalidNFTIndex)?;
```
need validate function

yymone commented on pallets/swap/src/lib.rs:
```diff
@@ -691,42 +698,126 @@ decl_module! {
+                return Err(Error::<T>::InvalidNFTIndex)?;
+            }
+
+            if recipient >= AccountIndexCount::get() || recipient == 0u32 {
```
better have another accountIndex validation function as this is the second time I see this check

yymone replied:
Better have a account index validation function as this is the second time I see such checking

yymone commented on pallets/swap/src/lib.rs:
```diff
@@ -691,42 +698,126 @@ decl_module! {
 
-            let req_id = req_id_get::<T>()?;
-            if recipent >= AccountIndexCount::get() {
+            if nft_id >= MAX_NFTINDEX_COUNT || nft_id == 0u32 {
```
same nft validation function

yymone replied:
need nft validation function

akayi07 commented on pallets/swap/src/aux.rs:
```diff
@@ -216,8 +224,44 @@ pub fn nft_transfer<T: Config>(
+) -> Result<(), Error<T>> {
+    let nft = NFTMap::get(&nft_id);
+    if nft.0 == 0u32 {
+        return Err(Error::<T>::NFTNoAuthority)?;
```
Yes, it's a mistake to add ? after it because the function return Result rather than DispatchResult.

yymone reviewed (commented):
Maybe need add some instruction in the main Readme about how to run the unit test.

yymone commented:
Good unit tests.

May need some improvements:

1. Need add error cases for catching error::<T>::InvalidSignature. Currently it cannot be achieved because currently you are using same public key for different accountIndex. In fact, we should use different public key for different account Index. Thus I suggest you can create pk0, pk1, pk2.json to represent different keys for different account.
2. We need make the unit tests reproducable and extensible. This means we need add the instruction in Readme to tell people how to run the unit test, and how to create new unit tests. This may include upload your own tools into ZKP NFT branch within a seperate directory like zkp/tests/ToosForSubstrateNodeUnitTests. And instructions for how to generate sign for new unit tests.

yymone commented on pallets/swap/src/tests.rs:
```diff
@@ -1,23 +1,2217 @@
+        assert_ok!(SwapModule::set_key(Origin::signed(origin), pub_key));
+
+        origin = 0u64;
+        let sign_deposit0: [u8; 64] = get_sign("deposit2");
```
should be sign_deposit2?

akayi07 replied:
Yes, it;s a mistake.

akayi07 replied:
No, it means we want to deposit accountIndex2, but the origin is not admin so it has no access. I will add comments to make it clear.

yymone replied:
sure, thanks.

akayi07 replied:
Yes, you are right. It shoude be "2".

yymone commented on pallets/swap/src/tests.rs:
```diff
@@ -1,23 +1,2217 @@
+        origin = 1u64;
+        let sign_deposit2: [u8; 64] = get_sign("deposit2");
+        let account_index = 2u32;
+        let token_index = 0u32;
```
For all invalid unit test cases, better add comments to specify which paramter is wrong and cause the issue.

For example, this one is very had to figure out as you just change the token_index from 1 to 0. And it is becuase the signature was generated by your tool with token_index=1.

This is very hard to figure out for a guy which come to check you unit test code if no comments.

akayi07 replied:
Sure. I will add comments.

yymone commented on pallets/swap/src/errors.rs:
```diff
@@ -33,6 +33,7 @@ decl_error! {
         L1AccountOverflow,
         InvalidTokenPair,
         InvalidTokenIndex,
+        InvalidNFTIndex,
```
Any new added enum must be added at last.
This is because maybe it will be changed to number and add any new enum in the middle will change the num index after it.
If any other code use number, it will be wrong.

akayi07 replied:
I understand. I will modify it.

yymone commented on pallets/swap/src/errors.rs:
```diff
@@ -37,5 +36,7 @@ decl_error! {
         InvalidKey,
         InvalidSignature,
         InvalidAccount
+        IsNotOwner,
```
Please check this.
In fact you delete the NFTNoAuthority and replace it with IsNotOwner.
Thus either you add back the NFTNoAuthority then add IsNotOwner at last, or just replace the NFTNoAuthority with IsNotOwner.
Current change is not good because it will reduce all the enum number after  original NFTNoAuthority.

akayi07 replied:
Yes, I delete the NFTNoAuthority because I do not see it in anywhere expect error.rs. Yes, I reduced all the enum number after NFTNoAuthority. I will add it back.

yymone commented on pallets/swap/src/errors.rs:
```diff
@@ -37,5 +36,7 @@ decl_error! {
         InvalidKey,
         InvalidSignature,
         InvalidAccount
```
miss "," here

akayi07 replied:
I'm sorry. I will compile and test after each change from now on.

akayi07 commented:
@yymone Hi, I have modified the code in error.rs. I'm sorry for not running 'cargo test' after I changed the code. I run cargo test this time:[
![Uploading 微信图片_20220422144410.png…]()
](url)
